src/sniffer.c

```diff
@@ -7132,6 +7191,308 @@ int ssl_PollSniffer(WOLF_EVENT** events, int maxEvents, WOLF_EVENT_FLAG flags,
+{
+    unsigned int hash = 0;
+    for (int i = 0; i < CLIENT_RANDOM_LENGTH; i++) {
+        hash = (hash * CLIENT_RANDOM_BITS + clientRandom[i]) % WOLFSSL_SNIFFER_KEYLOGFILE_HASH_TABLE_SIZE;
+    }
+    return hash;
+}
+
+static int addSecretNode(unsigned char* clientRandom, unsigned char* masterSecret, char* error)
+{
+    SecretNode* newSecretNode = (SecretNode*)XMALLOC(sizeof(SecretNode), NULL, DYNAMIC_TYPE_SNIFFER_KEYLOG_NODE);
+    if (newSecretNode == NULL) {
+        SetError(MEMORY_STR, error, NULL, 0);
+        return WOLFSSL_SNIFFER_ERROR;
+    }
+
+    XMEMCPY(newSecretNode->clientRandom, clientRandom, CLIENT_RANDOM_LENGTH);
+    XMEMCPY(newSecretNode->masterSecret, masterSecret, MASTER_SECRET_LENGTH);
+
+    LOCK_SECRET_LIST();
+
+    unsigned int index = secretHashFunction(clientRandom);
+    newSecretNode->next = NULL;
+
+    if (secretHashTable[index] == NULL) {
+        secretHashTable[index] = newSecretNode;
+    }
+    else {
+        SecretNode* current = secretHashTable[index];
+        while (current != NULL) {
+            if (memcmp(current->clientRandom, clientRandom, CLIENT_RANDOM_LENGTH) == 0) {
+                // BRN-sniffer-TODO: what if the same client random has a different master secret?
+                // Should we just update it? or return error?
+                fprintf(stderr, "Found duplicate client random value in keylog file. Rejecting\n");
+                XFREE(newSecretNode, NULL, DYNAMIC_TYPE_SNIFFER_KEYLOG_NODE);
+                break;
+            }
+            if (current->next == NULL) {
+                current->next = newSecretNode;
+                break;
+            }
+            current = current->next;
+        }
+    }
+
+    UNLOCK_SECRET_LIST();
+
+    return 0;
+}
+
+
+/*
+ * Looks up a master secret for a given client random from the keylog file
+ */
+static unsigned char* findMasterSecret(unsigned char* clientRandom)
+{
+    unsigned char* secret = NULL;
+
+    LOCK_SECRET_LIST();
+
+    unsigned int index = secretHashFunction(clientRandom);
+    SecretNode* node = secretHashTable[index];
+
+    while(node != NULL) {
+        if(XMEMCMP(node->clientRandom, clientRandom, CLIENT_RANDOM_LENGTH) == 0) {
+            secret = node->masterSecret;
+            break;
+        }
+        node = node->next;
+    }
+
+    UNLOCK_SECRET_LIST();
+
+    return secret;
+}
+
+
+static void hexToBin(const char* hex, unsigned char* bin, int binLength)
+{
+    for(int i = 0; i < binLength; i++) {
+        sscanf(hex + 2 * i, "%02hhx", &bin[i]);
+    }
+}
+
+static int parseKeyLogFile(const char* fileName, char* error)
+{
+    int ret = 0;
+
+    FILE* file = fopen(fileName, "r");
+    if (file == NULL) {
+        fprintf(stderr, "Could not open keylog file: %s\n", fileName);
+        SetError(KEYLOG_FILE_INVALID, error, NULL, 0);
+        return WOLFSSL_SNIFFER_ERROR;
+    }
+
+    const char CLIENT_RANDOM_LABEL_STR[] = "CLIENT_RANDOM";
```

Comment:
No mixed declarations. Keep all variable declarations at top of function or brace section.